src/clib/pio_gdal.c

```diff
@@ -0,0 +1,1350 @@
+    int nvars = 0;
+    int *rec_var = NULL;
+    int *pio_type = NULL;
+    int *pio_type_size = NULL;
+    MPI_Datatype *mpi_type = NULL;
+    int *mpi_type_size = NULL;
+    int *ndims = NULL;
+    int mpierr = MPI_SUCCESS, mpierr2;  /** Return code from MPI function codes. */
+    int ierr = PIO_NOERR;      /* Return code from function calls. */
+    GDALDatasetH hDS;
+
+#ifdef USE_MPE
+    pio_start_mpe_log(OPEN);
+#endif /* USE_MPE */
+
+    /* Get the IO system info from the iosysid. */
+    if (!(ios = pio_get_iosystem_from_id(iosysid)))
+        return pio_err(NULL, NULL, PIO_EBADID, __FILE__, __LINE__);
+
+    /* User must provide valid input for these parameters. */
+    if (!hDSp || !iotype || !filename)
+        return pio_err(ios, NULL, PIO_EINVAL, __FILE__, __LINE__);
+    if (*iotype != PIO_IOTYPE_GDAL )
+        return pio_err(ios, NULL, PIO_EINVAL, __FILE__, __LINE__);
+
+    /* Allocate space for the file info. */
+    if (!(file = calloc(sizeof(*file), 1)))
+        return pio_err(ios, NULL, PIO_ENOMEM, __FILE__, __LINE__);
+
+    /* Fill in some file values. */
+    file->fh = -1;
+    file->iotype = *iotype;
+    file->iosystem = ios;
+    file->writable = (mode & PIO_WRITE) ? 1 : 0;
+
+    /* Set to true if this task should participate in IO (only true
+     * for one task with netcdf serial files. */
+    if (file->iotype == PIO_IOTYPE_GDAL ||
+        ios->io_rank == 0)
+        file->do_io = 1;
+
+    /* If async is in use, and this is not an IO task, bcast the parameters. */
+    if (ios->async)
+    {
+        int msg = PIO_MSG_OPEN_FILE;
+        size_t len = strlen(filename);
+
+        if (!ios->ioproc)
+        {
+            /* Send the message to the message handler. */
+            if (ios->compmaster == MPI_ROOT)
+                mpierr = MPI_Send(&msg, 1, MPI_INT, ios->ioroot, 1, ios->union_comm);
+
+            /* Send the parameters of the function call. */
+            if (!mpierr)
+                mpierr = MPI_Bcast(&len, 1, MPI_INT, ios->compmaster, ios->intercomm);
+            if (!mpierr)
+                mpierr = MPI_Bcast((void *)filename, len + 1, MPI_CHAR, ios->compmaster, ios->intercomm);
+            if (!mpierr)
+                mpierr = MPI_Bcast(&file->iotype, 1, MPI_INT, ios->compmaster, ios->intercomm);
+            if (!mpierr)
+                mpierr = MPI_Bcast(&mode, 1, MPI_INT, ios->compmaster, ios->intercomm);
+        }
+
+        /* Handle MPI errors. */
+        if ((mpierr2 = MPI_Bcast(&mpierr, 1, MPI_INT, ios->comproot, ios->my_comm)))
+            return check_mpi(NULL, file, mpierr2, __FILE__, __LINE__);
+        if (mpierr)
+            return check_mpi(NULL, file, mpierr, __FILE__, __LINE__);
+    }
+
+    /* If this is an IO task, then call the netCDF function. */
+    if (ios->ioproc)
+    {
+        switch (file->iotype)
+        {
+        case PIO_IOTYPE_GDAL:
+//            if (ios->io_rank == 0)
```

Comment:
Why is this commented out?  If your IO is limited to a serial interface you only want to open the file on the root io task.